File: dclib/core/src/main/java/dclib/util/Maths.java
package dclib.util;

import com.badlogic.gdx.math.MathUtils;

public final class Maths {

    public static final float DEGREES_MAX = 360;
	public static final float RADIANS_MAX = MathUtils.PI2;

	private Maths() {
	}

	public static final boolean between(final float value, final float min, final float max) {
		return value >= min && value <= max;
	}

	public static final float min(final float currentValue, final float newValue) {
		return Float.isNaN(currentValue) ? newValue : Math.min(currentValue, newValue);
	}

	public static final float max(final float currentValue, final float newValue) {
		return Float.isNaN(currentValue) ? newValue : Math.max(currentValue, newValue);
	}

	public static final float distance(final float value1, final float value2) {
		return Math.abs(value1 - value2);
	}

	public static final float degDistance(final float deg1, final float deg2) {
        float phi = distance(deg1, deg2) % DEGREES_MAX;
        return phi > DEGREES_MAX / 2 ? DEGREES_MAX - phi : phi;
    }

	/**
	 * Rounds down to the nearest number that is divisible by the interval length.
	 * @param value value to round down from
	 * @param intervalLength length between each interval
	 * @return rounded down value
	 */
	public static final float round(final float value, final float intervalLength) {
		return value - value % intervalLength;
	}

}

File: dclib/core/src/main/java/dclib/util/Maths.kt
package dclib.util

import com.badlogic.gdx.math.Vector2
import dclib.geometry.VectorUtils

object Maths {
    val DEGREES_MAX = 360f
    val HALF_DEGREES_MAX = DEGREES_MAX / 2

    fun min(currentValue: Float, newValue: Float): Float {
        return if (java.lang.Float.isNaN(currentValue)) newValue else Math.min(currentValue, newValue)
    }

    fun max(currentValue: Float, newValue: Float): Float {
        return if (java.lang.Float.isNaN(currentValue)) newValue else Math.max(currentValue, newValue)
    }

    fun minAbs(value: Float, maxValue: Float): Float {
        return Math.min(Math.abs(value), maxValue) * Math.signum(value)
    }

    fun distance(value1: Float, value2: Float): Float {
        return Math.abs(value1 - value2)
    }

    fun degDelta(toDegrees: Float, fromDegrees: Float): Float {
        return ((toDegrees - fromDegrees + DEGREES_MAX + HALF_DEGREES_MAX) % DEGREES_MAX) - HALF_DEGREES_MAX
    }

    /**
     * Rounds down to the nearest integer that is divisible by the interval length.
     * @param value value to round down from
     * *
     * @param intervalLength length between each interval
     * *
     * @return rounded down value
     */
    fun round(value: Float, intervalLength: Float): Float {
        return value - value % intervalLength
    }

    fun getScaledRotation(degrees: Float, scale: Vector2): Float {
        val roundedDegrees = Maths.round(degrees, Maths.DEGREES_MAX)
        var scaledRemainderDegrees = VectorUtils.toVector2(degrees, 1f).scl(scale).angle()
        // Handle the case where the passed in degrees were negative, so they should be kept negative
        if (degrees < 0) {
            scaledRemainderDegrees -= Maths.DEGREES_MAX
        }
        return roundedDegrees + scaledRemainderDegrees
    }
}
